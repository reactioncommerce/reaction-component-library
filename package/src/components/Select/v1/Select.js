import React, { Component } from "react";
import isEqual from "lodash.isequal";
import PropTypes from "prop-types";
import ReactSelect from "react-select";
import { applyTheme, CustomPropTypes } from "../../../utils";

// This is currently mostly to ensure that this stays above our code examples in the
// style guide UI, which have zIndexes of 2 and 3 in some places. This might need to
// be taken from the theme eventually, though, if people have problems in other places.
const MENU_Z_INDEX = 4;

const nullDefaultEquals = (value1, value2) => ((value1 || null) === (value2 || null));

// Rather than pass through all props to react-select, we'll keep a whitelist
// to better control the usage and appearance of this component.
const supportedPassthroughProps = [
  "autoFocus",
  "backspaceRemovesValue",
  "blurInputOnSelect",
  "captureMenuScroll",
  "closeMenuOnSelect",
  "components",
  "escapeClearsValue",
  "formatGroupLabel",
  "formatOptionLabel",
  "getOptionLabel",
  "getOptionValue",
  "hideSelectedOptions",
  "inputValue",
  "isClearable",
  "isLoading",
  "isOptionDisabled",
  "isRtl",
  "isSearchable",
  "loadingMessage",
  "minMenuHeight",
  "maxMenuHeight",
  "maxValueHeight",
  "menuIsOpen",
  "menuPlacement",
  "noOptionsMessage",
  "onBlur",
  "onFocus",
  "onInputChange",
  "onKeyDown",
  "onMenuOpen",
  "onMenuClose",
  "onMenuScrollToTop",
  "onMenuScrollToBottom",
  "pageSize",
  "placeholder",
  "screenReaderStatus",
  "scrollMenuIntoView",
  "tabSelectsValue",
  "value"
];

function applyValidationColor(themeProp = "color") {
  return (props) => {
    let status;
    if (props.errors && props.errors.length) {
      status = "error";
    } else if (props.hasBeenValidated && props.value && props.value.length) {
      status = "success";
    } else if (props.isFocused) {
      status = "focus";
    } else {
      status = "default";
    }
    return applyTheme(`${themeProp}_${status}`)(props);
  };
}

const getInputBorderColor = applyValidationColor("inputBorderColor");
const getInputBorderRadius = applyTheme("inputBorderRadius");

function getCustomStyles(props) {
  const { maxWidth } = props;

  // TODO isDark change bg color
  return {
    container(base) {
      return { ...base, maxWidth };
    },
    control(base, state) {
      return {
        ...base,
        borderColor: getInputBorderColor({ ...props, isFocused: state.isFocused }),
        borderRadius: getInputBorderRadius(props)
      };
    },
    menu(base) {
      return {
        ...base,
        zIndex: MENU_Z_INDEX
      };
    }
  };
}

class Select extends Component {
  static propTypes = {
    /**
     * Passed through to react-select package. Focus the control when it is mounted
     */
    autoFocus: PropTypes.bool,

    /**
     * Passed through to react-select package. Remove the currently focused option when the user presses backspace
     */
    backspaceRemovesValue: PropTypes.bool,

    /**
     * Passed through to react-select package.
     * Remove focus from the input when the user selects an option (handy for dismissing the keyboard on touch devices)
     */
    blurInputOnSelect: PropTypes.bool,

    /**
     * Passed through to react-select package. When the user reaches the top/bottom of the menu, prevent scroll on the scroll-parent
     */
    captureMenuScroll: PropTypes.bool,

    /**
     * Passed through to react-select package. Close the select menu when the user selects an option
     */
    closeMenuOnSelect: PropTypes.bool,

    /**
     * Passed through to react-select package. Custom components to use
     */
    components: PropTypes.object,

    /**
     * An array of validation errors
     */
    errors: PropTypes.array,

    /**
     * Passed through to react-select package. Clear all values when the user presses escape AND the menu is closed
     */
    escapeClearsValue: PropTypes.bool,

    /**
     * Passed through to react-select package. Formats group labels in the menu as React components
     */
    formatGroupLabel: PropTypes.func,

    /**
     * Passed through to react-select package. Formats option labels in the menu and control as React components
     */
    formatOptionLabel: PropTypes.func,

    /**
     * Passed through to react-select package. Resolves option data to a string to be displayed as the label by components
     */
    getOptionLabel: PropTypes.func,

    /**
     * Passed through to react-select package. Resolves option data to a string to compare options and specify value attributes
     */
    getOptionValue: PropTypes.func,

    /**
     * Enable when a input's value has been validated
     */
    hasBeenValidated: PropTypes.bool,

    /**
     * Passed through to react-select package. Hide the selected option from the menu
     */
    hideSelectedOptions: PropTypes.bool,

    /**
     * Passed through to react-select package. The value of the search input
     */
    inputValue: PropTypes.string,

    /**
     * Passed through to react-select package. Is the select value clearable
     */
    isClearable: PropTypes.bool,

    /**
     * Passed through to react-select package. Is the select in a state of loading (async)
     */
    isLoading: PropTypes.bool,

    /**
     * Passed through to react-select package. Override the built-in logic to detect whether an option is disabled
     */
    isOptionDisabled: PropTypes.func,

    /**
     * Passed through to react-select package as `isDisabled`. Should the user be able to edit this value?
     */
    isReadOnly: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),

    /**
     * Passed through to react-select package. Is the select direction right-to-left
     */
    isRtl: PropTypes.bool,

    /**
     * Passed through to react-select package. Whether to enable search functionality
     */
    isSearchable: PropTypes.bool,

    /**
     * Passed through to react-select package. Async: Text to display when loading options
     */
    loadingMessage: PropTypes.func,

    /**
     * Passed through to react-select package. Maximum height of the menu before scrolling
     */
    maxMenuHeight: PropTypes.number,

    /**
     * Passed through to react-select package. Maximum height of the value container before scrolling
     */
    maxValueHeight: PropTypes.number,

    /**
     * Passed through to react-select package. Whether the menu is open
     */
    menuIsOpen: PropTypes.bool,

    /**
     * Passed through to react-select package. Default placement of the menu in relation to the control. 'auto' will flip
     * when there isn't enough space below the control.
     */
    menuPlacement: PropTypes.oneOf(["auto", "bottom", "top"]),

    /**
     * Passed through to react-select package. Minimum height of the menu before flipping
     */
    minMenuHeight: PropTypes.number,

    /**
     * A name or object path that determines where in the closest form object this will appear.
     */
    name: PropTypes.string,

    /**
     * Passed through to react-select package. Text to display when there are no options
     */
    noOptionsMessage: PropTypes.func,

    /**
     * Passed through to react-select package. Handle blur events on the control
     */
    onBlur: PropTypes.func,

    /**
     * Called with the new selected value each time the user changes the selection
     */
    onChange: PropTypes.func,

    /**
     * Called with the new selected value each time the user changes the selection
     */
    onChanging: PropTypes.func,

    /**
     * Passed through to react-select package. Handle focus events on the control
     */
    onFocus: PropTypes.func,

    /**
     * Passed through to react-select package. Handle change events on the input
     */
    onInputChange: PropTypes.func,

    /**
     * Passed through to react-select package. Handle key down events on the select
     */
    onKeyDown: PropTypes.func,

    /**
     * Passed through to react-select package. Handle the menu closing
     */
    onMenuClose: PropTypes.func,

    /**
     * Passed through to react-select package. Handle the menu opening
     */
    onMenuOpen: PropTypes.func,

    /**
     * Passed through to react-select package. Fired when the user scrolls to the bottom of the menu
     */
    onMenuScrollToBottom: PropTypes.func,

    /**
     * Passed through to react-select package. Fired when the user scrolls to the top of the menu
     */
    onMenuScrollToTop: PropTypes.func,

    /**
     * The options to show, in Composable Form Spec format.
     * @see http://forms.dairystatedesigns.com/user/input/#selection-inputs
     */
    options: CustomPropTypes.options,

    /**
     * Passed through to react-select package. Number of options to jump in menu when page{up|down} keys are used
     */
    pageSize: PropTypes.number,

    /**
     * Passed through to react-select package. Placeholder text for the select value
     */
    placeholder: PropTypes.string,

    /**
     * Passed through to react-select package. Status to relay to screen readers
     */
    screenReaderStatus: PropTypes.func,

    /**
     * Passed through to react-select package. Whether the menu should be scrolled into view when it opens
     */
    scrollMenuIntoView: PropTypes.bool,

    /**
     * Passed through to react-select package. Select the currently focused option when the user presses tab
     */
    tabSelectsValue: PropTypes.bool,

    /**
     * Set this to the current saved value, if editing, or a default value if creating. The closest form implementing
     * the Composable Forms spec will pass this automatically.
     */
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool
    ])
  };

  static defaultProps = {
    isReadOnly: false,
    isSearchable: false,
    onChange() {},
    onChanging() {},
    options: []
  };

  static isFormInput = true;

  constructor(props) {
    super(props);

    this.validateOptions(props.options);

    this.state = {
      value: props.value || null
    };
  }

  componentWillMount() {
    this.handleChanged(this.state.value);
  }

  componentWillReceiveProps(nextProps) {
    const { options, value } = this.props;
    const { options: nextOptions, value: nextValue } = nextProps;

    // Whenever a changed value prop comes in, we reset state to that, thus becoming clean.
    if (!nullDefaultEquals(value, nextValue)) {
      this.setValue(nextValue);
    }

    if (!isEqual(options, nextOptions)) {
      this.validateOptions(nextOptions);
    }
  }

  getValue() {
    return this.state.value;
  }

  setValue(value) {
    this.setState({ value });
    this.handleChanged(value);
  }

  resetValue() {
    this.setValue(this.props.value);
  }

  handleChanged = (value) => {
    const { onChange, onChanging } = this.props;
    if (value !== this.lastValue) {
      this.lastValue = value;
      onChanging(value);
      onChange(value);
    }
  };

  handleSelectLibChanged = (selection) => {
    let { value } = selection || {};

    if (value !== undefined && value !== null) {
      switch (this.dataType) {
        case "string":
          value = String(value);
          break;
        case "number":
          value = value === "" ? null : Number(value);
          break;
        case "boolean":
          value = value === "" ? null : Boolean(value);
          break;
        default:
          // do nothing
      }
    }

    this.setValue(value || null);
  };

  // Input is dirty if value prop doesn't match value state. Whenever a changed
  // value prop comes in, we reset state to that, thus becoming clean.
  isDirty() {
    return (this.state.value || null) !== (this.props.value || null);
  }

  // Make sure all option values have the same data type, and record what that is
  validateOptions(options) {
    (options || []).forEach((option) => {
      if (option.optgroup) {
        this.validateOptions(option.options);
      } else {
        const checkDataType = typeof option.value;
        if (!this.dataType) {
          this.dataType = checkDataType;
        } else if (checkDataType !== this.dataType) {
          throw new Error(`All option values must have the same data type. The data type of the first option is "${this.dataType}" while the` +
            ` data type of the ${option.label} option is "${checkDataType}"`);
        }
      }
    });
  }

  render() {
    const { isReadOnly, options } = this.props;
    const { value } = this.state;

    // Unfortunately right now, react-select optgroup support is just a tad different from the
    // composable form spec. Might be able to do a PR to get react-select updated.
    const reactSelectOptions = options.map((opt) => {
      if (opt.options) {
        return {
          label: opt.optgroup,
          options: opt.options
        };
      }
      return opt;
    });

    const passthroughProps = {};
    supportedPassthroughProps.forEach((prop) => {
      passthroughProps[prop] = this.props[prop];
    });

    let optionValue;
    if (value !== undefined && value !== null) {
      optionValue = reactSelectOptions.find((opt) => {
        if (opt.options) return opt.options.find((o) => o.value === value);
        return opt.value === value;
      });
    }

    return (
      <ReactSelect
        {...passthroughProps}
        isDisabled={isReadOnly}
        value={optionValue}
        onChange={this.handleSelectLibChanged}
        options={reactSelectOptions}
        styles={getCustomStyles(this.props)}
      />
    );
  }
}

export default Select;
