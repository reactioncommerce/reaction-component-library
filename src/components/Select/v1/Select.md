### Select Overview

Selects are form elements that allow the user to choose a single value from a set of options.

#### Choosing a Select style

There are two basic types of selects: simple and searchable.

The simple select can be used in cases where there are fewer than 10 options.

```jsx
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<Select options={options} />
```

A searchable select should be used when there are more than 10 options or when the options are not known until the user starts typing.

```jsx
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<Select options={options} isSearchable />
```

In cases where you want a smaller width for the select, set `maxWidth` property to a number of pixels based on how much space your longest option label needs.

```jsx
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<Select options={options} maxWidth={150} />
```

#### States

A select can be in one of three states: normal, invalid, or valid. Normal state is as shown previously

##### Invalid State

When used within a form, a selected value might be deemed invalid, either because the user has not selected a value or because the user has selected a value that is not allowed based on other information entered in the same form. In this case, one or more error objects can be passed in the `errors` prop and will cause the select to appear as invalid.

```jsx
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<Select name="flavor" options={options} errors={[{ name: "flavor", message: "Please choose one" }]} />
```

##### Valid State

When used within a form, a selected value might be deemed valid after its value has been checked. If the `errors` prop is empty and the `hasBeenValidated` prop is true and there is a selected value, the select will appear valid.

```jsx
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<Select options={options} value="vanilla" hasBeenValidated />
```
