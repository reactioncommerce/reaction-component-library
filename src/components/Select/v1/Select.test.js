import React from "react";
import renderer from "react-test-renderer";
import Select from "./Select";

test("basic snapshot", () => {
  const component = renderer.create(<Select />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
