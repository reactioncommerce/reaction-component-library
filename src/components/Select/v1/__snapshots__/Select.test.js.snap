// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
<div
  className="react-select__container css-zckpcq"
  onKeyDown={[Function]}
>
  <span
    aria-atomic="true"
    aria-live="polite"
    className="css-17l7ekz"
    role="status"
  >
    0 results available.
  </span>
  <div
    className="react-select__control css-1lc6nq7"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="react-select__value-container css-6rkci4"
    >
      <div
        className="react-select__placeholder css-16ucli7"
      >
        Select...
      </div>
      <input
        className="css-hg63qv"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        readOnly={true}
        value=""
      />
    </div>
    <div
      className="react-select__indicators css-1lfa1d8"
    >
      <span
        className="react-select__indicator-separator css-1cnd6ox"
        css={undefined}
        role="presentation"
      />
      <div
        className="react-select__indicator react-select__dropdown-indicator css-1c0gw5t"
        css={undefined}
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        role="button"
      >
        <svg
          className="react-select__icon react-select__down-icon css-qyb9b"
          focusable="false"
          height={20}
          role="presentation"
          viewBox="0 0 20 20"
          width={20}
        >
          <path
            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
